app/controllers/automated_tests_controller.rb

```diff
@@ -35,14 +35,26 @@ def student_interface
     @student = current_role
     @grouping = @student.accepted_grouping_for(@assignment.id)
 
-    unless @grouping.nil?
+    if @grouping.present?
       @grouping.refresh_test_tokens
       @authorized = flash_allowance(:notice,
                                     allowance_to(:run_tests?,
                                                  current_role,
                                                  context: { assignment: @assignment, grouping: @grouping })).value
-    end
 
+      if @assignment.enable_student_tests
+        if @assignment.non_regenerating_tokens
+          @next_token_generation_time = nil
+        else
+          hours_from_start = [(Time.current - @assignment.token_start_date) / 3600, 0].max
+          periods_from_start = (hours_from_start / @assignment.token_period).floor
+          last_period_begin = @assignment.token_start_date + (periods_from_start * @assignment.token_period).hours
+          @next_token_generation_time = last_period_begin + @assignment.token_period.hours
+        end
+        @next_token_generation_time = I18n.l(@next_token_generation_time) if @next_token_generation_time
```

Comment:
This line of code should be merged into the else branch above, and you don't need the `if @next_token_generation_time`. That is, if the else branch executes then we need to call `I18n.l`, otherwise we don't.